### DISCUSSION TITLE
Wrong cart and line items cost - total and subtotal

### DISCUSSION THREAD

**Discussion by hagsonz** [Shopify/function-examples#430] [Feedback: Discounts APIs]
Using product discount. for the cart, I am getting same amount for the total and subtotal fields. in addition, tax and duty values are null.
also, if i have a coupon applied, the Input is still the exact same. 

for the line items, the item which is discounted by the function is with full price which is logical as the function haven't applied the discount yet. but the one which is discounted by a coupon, is still with full price.

how can i calculate the correct amount of the cart and line items?

<img width="692" alt="image" src="https://github.com/Shopify/function-examples/assets/28609586/f6bc6d92-a846-44f8-b326-ca2bc9f9a715">

<img width="613" alt="image" src="https://github.com/Shopify/function-examples/assets/28609586/40e7440c-f666-40c5-a9a2-86d4bdbb026c">

**Input (STDIN)**

``` 
{
  "presentmentCurrencyRate": "1.0",
  "cart": {
    "cost": {
      "subtotalAmount": {
        "amount": "1500.0",
        "currencyCode": "ILS"
      },
      "totalAmount": {
        "amount": "1500.0",
        "currencyCode": "ILS"
      },
      "totalTaxAmount": null,
      "totalDutyAmount": null
    },
    "lines": [
      {
        "id": "gid://shopify/CartLine/0",
        "quantity": 1,
        "cost": {
          "totalAmount": {
            "amount": "600.0",
            "currencyCode": "ILS"
          },
          "subtotalAmount": {
            "amount": "600.0",
            "currencyCode": "ILS"
          }
        },
        "merchandise": {
          "__typename": "ProductVariant",
          "id": "gid://shopify/ProductVariant/44579562455318",
          "product": {
            "id": "gid://shopify/Product/8218915930390"
          }
        },
        "attribute": null
      },
      {
        "id": "gid://shopify/CartLine/1",
        "quantity": 1,
        "cost": {
          "totalAmount": {
            "amount": "900.0",
            "currencyCode": "ILS"
          },
          "subtotalAmount": {
            "amount": "900.0",
            "currencyCode": "ILS"
          }
        },
        "merchandise": {
          "__typename": "ProductVariant",
          "id": "gid://shopify/ProductVariant/44579562881302",
          "product": {
            "id": "gid://shopify/Product/8218916258070"
          }
        },
        "attribute": {
          "key": "_order_bump_offer_id",
          "value": "7MWcCsmycOQbf4kZZStQx"
        }
      }
    ]
  },
  "discountNode": {
    "metafield": {
      "value": "{\"offers\":[{\"expanded\":true,\"enabled\":true,\"offerName\":\"Edit this offer name\",\"layout\":\"vertical\",\"cardLayout\":\"vertical\",\"minCartQuantity\":\"1\",\"maxCartQuantity\":\"601\",\"minCartValue\":0,\"maxCartValue\":\"1000\",\"products\":[\"gid://shopify/Product/8218916258070\"],\"onlyProducts\":[],\"discountAmount\":\"10\",\"discountType\":\"percentage\",\"cartMessage\":\"Edit me\",\"offerType\":\"product\",\"id\":\"7MWcCsmycOQbf4kZZStQx\"},{\"expanded\":true,\"enabled\":false,\"offerName\":\"Edit this offer name\",\"layout\":\"vertical\",\"cardLayout\":\"horizontal\",\"minCartQuantity\":0,\"maxCartQuantity\":99999,\"minCartValue\":0,\"maxCartValue\":100000,\"products\":[\"gid://shopify/Product/8218916126998\",\"gid://shopify/Product/8218916192534\",\"gid://shopify/Product/8218915963158\"],\"onlyProducts\":[],\"discountAmount\":\"5\",\"discountType\":\"percentage\",\"cartMessage\":\"Edit me\",\"offerType\":\"product\",\"id\":\"1wpsj73LB5LO_Ir9dSNzE\"}]}"
    }
  }
}

**Comment by thosakwe** [comment]
I noticed that for order discount functions, `cart.cost.totalAmount` and `cart.cost.subtotalAmount` are also equal.

**Comment by blacklen** [comment]
I noticed that the order discount created from the native Shopify function can be used properly with total (without other discount values) but it's not the same for the function. On the order discount function, totalAmount = subtotalAmount. Seem like It's not running after the subtotal is calculated as example in the screenshot below.
![Screenshot 2024-01-26 at 12 12 28](https://github.com/Shopify/function-examples/assets/31103446/366f5f78-b6ce-4fe0-9f12-669280960636)

**Comment by bkspace** [comment]
Have created a forum bug report for this: https://community.shopify.com/c/shopify-functions/wrong-cart-and-product-totals-in-order-discounts/m-p/2445374#M1546

**Comment by nickwesselman** [comment]
Product and order discount Functions are ran all at once, and their outputs are used later in the discounts engine, where, when determining the best discount combination, product discounts are considered first, then order discounts.

It is unfortunately not a bug, as discount Functions can’t know about other discounts when they’re being executed, as it’s long before the discounts engine begins to consider discount combinations and the best one. We’re going to update the documentation to make this clearer.

[Discount allocators](https://shopify.dev/docs/apps/selling-strategies/discounts/discount-function-examples/discounts-allocator) (currently in dev preview) would allow you to provide more control over the combination of order and product discounts.

**Reply by bkspace** [reply]
Thanks Nick! Would love to offer this in Discount Kit, but the documentation suggests that allocators are just for custom apps

**Reply by beverlytang** [reply]
Then what is the point of having subtotalAmount and totalAmount if they will always return the price before discount?
